src/Sylius/Component/Core/StateResolver/OrderPaymentStateResolver.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\StateResolver;
+
+use SM\Factory\FactoryInterface;
+use SM\StateMachine\StateMachineInterface;
+use Sylius\Component\Order\Model\OrderInterface;
+use Sylius\Component\Order\StateResolver\StateResolverInterface;
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\OrderPaymentStates;
+use Sylius\Component\Core\OrderPaymentTransitions;
+
+/**
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ * @author Arkadiusz Krakowiak <arkadiusz.krakowiak@lakion.com>
+ * @author Grzegorz Sadowski <grzegorz.sadowski@lakion.com>
+ */
+class OrderPaymentStateResolver implements StateResolverInterface
+{
+    /**
+     * @var FactoryInterface
+     */
+    private $stateMachineFactory;
+
+    /**
+     * @param FactoryInterface $stateMachineFactory
+     */
+    public function __construct(FactoryInterface $stateMachineFactory)
+    {
+        $this->stateMachineFactory = $stateMachineFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function resolve(OrderInterface $order)
+    {
+        $stateMachine = $this->stateMachineFactory->get($order, OrderPaymentTransitions::GRAPH);
+
+        if (OrderPaymentStates::STATE_PAID === $order->getPaymentState()) {
+            return;
+        }
+
+        if ($order->hasPayments()) {
+            $completedPaymentTotal = 0;
+            $payments = $order->getPayments()->filter(function (PaymentInterface $payment) {
```

Comment:
What about adding optional argument to ``$order->getPayment()`` function (as its done with [adjustments](https://github.com/Sylius/Sylius/blob/master/src/Sylius/Component/Order/Model/Order.php#L326))?